packages/concerto-core/lib/introspect/mapdeclaration.js

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+const MapPropertyType = require('./mapvaluetype');
+const MapKeyType = require('./mapkeytype');
+const ModelUtil = require('../modelutil');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapDeclaration defines an MapType of Key & Value pair.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapDeclaration extends Decorated {
+    /**
+     * Create an MapDeclaration.
+     * @param {ModelFile} modelFile the ModelFile for this class
+     * @param {Object} ast - The AST created by the parser
+     * @throws {IllegalModelException}
+     */
+    constructor(modelFile, ast) {
+        super(ast);
+        this.modelFile = modelFile;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+
+        if (this.ast.properties.length !== 2) {
+            throw new IllegalModelException(`MapDeclaration must contain exactly two properties -  MapKeyType & MapyValueType ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
+
+        const k = this.ast.properties.find(p => p.$class === 'concerto.metamodel@1.0.0.MapKeyType');
+        const v = this.ast.properties.find(p => p.$class === 'concerto.metamodel@1.0.0.AggregateValueType' || p.$class === 'concerto.metamodel@1.0.0.AggregateRelationshipValueType');
+
+        if (!k) {
+            throw new IllegalModelException(`MapDeclaration must contain MapKeyType  ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
+
+        if (!v) {
+            throw new IllegalModelException(`MapDeclaration must contain AggregateValueType  ${this.ast.name}`, this.modelFile, this.ast.location);
+        }
```

Comment:
Nit, please use descriptive names for variables. `key`, `value`, `property` etc.